### DISCUSSION TITLE
More CurlDownloader retry cases?

### DISCUSSION THREAD

**Discussion by dzuelke** [composer/composer#10887] [General]
We've recently run into random CI failures where jobs on CircleCI machine instances, executed inside Docker in there (because they're Cloud Native Buildpack builds), often encounter random network timeouts:

```
 !     > In CurlDownloader.php line 377:
 !     >                                                                                
 !     >   curl error 28 while downloading https://lang-php.s3.us-east-1.amazonaws.com/  
 !     >   /dist-heroku-20-stable/packages.json: Connection timed out after 10003 mill  
 !     >   iseconds
```

Restart the job, and it works fine.

Obviously, the usual advice here would be "fix your network", but this being Docker inside VMs on a service with ephemeral instances itself running on IaaS... you get the idea.

IMO, the behavior here should be to:

1. retry on this type of error, and
2. possibly start with a lower value for the [currently hard-coded ten seconds](https://github.com/composer/composer/blob/3e844cc48fb8bdbbdd55a06b7a59312b0cafe764/src/Composer/Util/Http/CurlDownloader.php#L202) for connect timeouts (increasing the value with every retry), and
3. exposing more information in the exception message, e.g. print `CURLINFO_NAMELOOKUP_TIME` and `CURLINFO_CONNECT_TIME` from the progress info (so people can understand whether it was a DNS resolution or a connection timeout), and
4. possibly changing `restartWithDelay` to use an exponential backoff algorithm for its retry delay (something like `5*2**$retryNum`).

Y'all fine with me making a PR for some or all of the above?

**Comment by Seldaek** [comment]
1. Sure. Except I am not sure about connection timeout, if it failed in 10seconds is it really going to work again? https://github.com/composer/composer/blob/3e844cc48fb8bdbbdd55a06b7a59312b0cafe764/src/Composer/Util/Http/CurlDownloader.php#L372 handles it somehow but I guess if you use a lower timeout first then it makes sense to retry on timeout, and then only print the warning at line 374 if it failed more than maxRetries?
2. Sure.
3. Sure
4. it kinda is already.. I mean it's right now waiting: 0ms, 100ms, 500ms, then it fails hard after 3 failed retries. I don't know if we really want to do more retrying?

**Reply by dzuelke** [reply]
So the important bit about the line you quoted is that that only shows an additional warning about being offline; it does not trigger a retry. The heuristic is to detect that is checking if `namelookup_time` is 0, which I assume happens on one or more OSes when network interfaces are down and I guess `gethostbyname()` then returns immediately?

Yes, I was thinking to either print this warning at the end, or treat this case as a "fallthrough" one where no retry is made (because 'namelookup_time` is zero).

The thing about these transient connection errors is that sometimes, name lookups fail, or time out. Retrying immediately works immediately. Remember, DNS works over UDP, so it's absolutely not uncommon for a name lookup to hit this timeout because of a glitch somewhere along the network path.

From anecdotal evidence I've seen myself and in tickets for basically any library, language, etc, this somehow is compounded in Docker.

This can also happen a lot on Alpine, where the musl DNS resolver has known issues (https://github.com/nodejs/node/issues/15780, https://github.com/gliderlabs/docker-alpine/blob/master/docs/caveats.md), or Kubernetes (https://pracucci.com/kubernetes-dns-resolution-ndots-options-and-why-it-may-affect-application-performances.html).

**Reply by naderman** [reply]
Definitely seen this type of intermittent dns issues elsewhere, and indeed the length of the timeout has nothing to do with the likelihood of it working if you retry. So 10 seconds for a first connection timeout seems long, I'd probably rather go for something like 3 seconds and then retry with 10 second timeout?
